File: matlab/fft_validation.m
% filepath: /Users/P2956632/Documents/CS 3120/Music_Feature_Analysis/matlab/fft_validation.m
% FFT Validation for Music Feature Analysis
% Validates FFT-based feature extraction and Parseval's theorem

function fft_validation()
    fprintf('======================================================================\n');
    fprintf('FFT VALIDATION - Music Feature Analysis\n');
    fprintf('======================================================================\n\n');
    
    % Test 1: Parseval's Theorem
    test_parsevals_theorem();
    
    % Test 2: Spectral Centroid
    test_spectral_centroid();
    
    % Test 3: Spectral Rolloff
    test_spectral_rolloff();
    
    % Test 4: Windowing Effects
    test_windowing_effects();
    
    fprintf('\n======================================================================\n');
    fprintf('FFT VALIDATION COMPLETE\n');
    fprintf('======================================================================\n');
end

function test_parsevals_theorem()
    fprintf('Test 1: Parseval''s Theorem Validation\n');
    fprintf('----------------------------------------------------------------------\n');
    
    % Generate test signal
    fs = 44100;  % Sample rate
    duration = 1;
    t = 0:1/fs:duration-1/fs;
    test_signal = sin(2*pi*440*t) + 0.5*sin(2*pi*880*t);
    
    % Time-domain energy
    time_energy = sum(test_signal.^2);
    
    % Frequency-domain energy
    fft_mag = abs(fft(test_signal));
    freq_energy = sum(fft_mag.^2) / length(test_signal);
    
    % Calculate error
    error = abs(time_energy - freq_energy) / time_energy * 100;
    
    fprintf('  Time-domain energy:      %.6f\n', time_energy);
    fprintf('  Frequency-domain energy: %.6f\n', freq_energy);
    fprintf('  Error:                   %.6f%%\n', error);
    
    if error < 1
        fprintf('  Result: PASSED (error < 1%%)\n');
    else
        fprintf('  Result: FAILED (error >= 1%%)\n');
    end
    fprintf('\n');
end

function test_spectral_centroid()
    fprintf('Test 2: Spectral Centroid Computation\n');
    fprintf('----------------------------------------------------------------------\n');
    
    fs = 44100;
    duration = 1;
    t = 0:1/fs:duration-1/fs;
    
    % Low-frequency signal (100 Hz)
    low_freq_signal = sin(2*pi*100*t);
    
    % Compute FFT
    fft_mag = abs(fft(low_freq_signal));
    freqs = (0:length(fft_mag)-1) * fs / length(fft_mag);
    
    % Positive frequencies only
    half_idx = floor(length(fft_mag)/2);
    freqs_pos = freqs(1:half_idx);
    fft_mag_pos = fft_mag(1:half_idx);
    
    % Compute spectral centroid
    centroid = sum(fft_mag_pos .* freqs_pos') / sum(fft_mag_pos);
    
    fprintf('  Expected centroid: ~100 Hz\n');
    fprintf('  Computed centroid: %.2f Hz\n', centroid);
    
    if centroid >= 90 && centroid <= 110
        fprintf('  Result: PASSED (within ±10 Hz)\n');
    else
        fprintf('  Result: FAILED (outside expected range)\n');
    end
    fprintf('\n');
end

function test_spectral_rolloff()
    fprintf('Test 3: Spectral Rolloff (95th percentile)\n');
    fprintf('----------------------------------------------------------------------\n');
    
    fs = 44100;
    duration = 1;
    t = 0:1/fs:duration-1/fs;
    
    % Test signal (1000 Hz)
    test_signal = sin(2*pi*1000*t);
    
    % Compute FFT
    fft_mag = abs(fft(test_signal));
    freqs = (0:length(fft_mag)-1) * fs / length(fft_mag);
    
    % Positive frequencies
    half_idx = floor(length(fft_mag)/2);
    freqs_pos = freqs(1:half_idx);
    fft_mag_pos = fft_mag(1:half_idx);
    
    % Cumulative sum of power
    cumsum_power = cumsum(fft_mag_pos);
    total_power = cumsum_power(end);
    
    % Find 95th percentile
    rolloff_idx = find(cumsum_power >= 0.95 * total_power, 1, 'first');
    rolloff_freq = freqs_pos(rolloff_idx);
    
    fprintf('  Expected rolloff: ~1000 Hz\n');
    fprintf('  Computed rolloff: %.2f Hz\n', rolloff_freq);
    
    if rolloff_freq >= 900
        fprintf('  Result: PASSED (rolloff > 900 Hz)\n');
    else
        fprintf('  Result: FAILED (rolloff too low)\n');
    end
    fprintf('\n');
end

function test_windowing_effects()
    fprintf('Test 4: Windowing Effects (Hann vs Rectangular)\n');
    fprintf('----------------------------------------------------------------------\n');
    
    fs = 44100;
    duration = 1;
    t = 0:1/fs:duration-1/fs;
    
    % Two sine waves
    test_signal = sin(2*pi*440*t) + 0.5*sin(2*pi*880*t);
    
    % Rectangular window (no window)
    fft_rect = abs(fft(test_signal));
    
    % Hann window
    window_hann = hann(length(test_signal))';
    fft_hann = abs(fft(test_signal .* window_hann));
    
    % Find peaks
    half_idx = floor(length(fft_rect)/2);
    
    [~, peaks_rect] = findpeaks(fft_rect(1:half_idx), 'MinPeakHeight', max(fft_rect(1:half_idx))*0.1);
    [~, peaks_hann] = findpeaks(fft_hann(1:half_idx), 'MinPeakHeight', max(fft_hann(1:half_idx))*0.1);
    
    fprintf('  Peaks found (rectangular): %d\n', length(peaks_rect));
    fprintf('  Peaks found (Hann):        %d\n', length(peaks_hann));
    
    if length(peaks_rect) >= 2 && length(peaks_hann) >= 2
        fprintf('  Result: PASSED (both windows detected peaks)\n');
    else
        fprintf('  Result: FAILED (insufficient peaks detected)\n');
    end
    fprintf('\n');
end

File: matlab/fft_validation.py
"""
FFT Validation (Python Alternative to MATLAB)
Validates FFT-based feature extraction and Parseval's theorem.
"""

import numpy as np
from scipy import signal
import matplotlib.pyplot as plt

def validate_parsevals_theorem():
    """Test Parseval's theorem: time-domain energy == frequency-domain energy."""
    print("=" * 70)
    print("FFT VALIDATION - Parseval's Theorem")
    print("=" * 70)
    
    # Generate test signal
    fs = 44100  # Sample rate
    duration = 1
    t = np.arange(0, duration, 1/fs)
    test_signal = np.sin(2 * np.pi * 440 * t) + 0.5 * np.sin(2 * np.pi * 880 * t)
    
    # Time-domain energy
    time_energy = np.sum(test_signal**2)
    
    # Frequency-domain energy
    fft_mag = np.abs(np.fft.fft(test_signal))
    freq_energy = np.sum(fft_mag**2) / len(test_signal)
    
    # Calculate error
    error = abs(time_energy - freq_energy) / time_energy * 100
    
    print(f"\nTime-domain energy:   {time_energy:.6f}")
    print(f"Frequency-domain energy: {freq_energy:.6f}")
    print(f"Error:               {error:.6f}%")
    
    if error < 1:
        print("Result: PASSED (error < 1%)")
    else:
        print("Result: FAILED (error >= 1%)")
    
    return error < 1


def validate_spectral_centroid():
    """Validate spectral centroid computation."""
    print("\n" + "=" * 70)
    print("SPECTRAL CENTROID VALIDATION")
    print("=" * 70)
    
    fs = 44100
    duration = 1
    t = np.arange(0, duration, 1/fs)
    
    # Low-frequency signal (100 Hz)
    low_freq_signal = np.sin(2 * np.pi * 100 * t)
    
    # Compute FFT
    fft_mag = np.abs(np.fft.fft(low_freq_signal))
    freqs = np.fft.fftfreq(len(low_freq_signal), 1/fs)
    
    # Positive frequencies only
    pos_mask = freqs >= 0
    freqs_pos = freqs[pos_mask]
    fft_mag_pos = fft_mag[pos_mask]
    
    # Compute spectral centroid
    centroid = np.sum(fft_mag_pos * freqs_pos) / np.sum(fft_mag_pos)
    
    print(f"\nExpected centroid: ~100 Hz")
    print(f"Computed centroid: {centroid:.2f} Hz")
    
    passed = 90 <= centroid <= 110
    if passed:
        print("Result: PASSED (within ±10 Hz)")
    else:
        print("Result: FAILED (outside expected range)")
    
    return passed


def visualize_fft_analysis():
    """Create FFT analysis visualizations."""
    print("\n" + "=" * 70)
    print("FFT ANALYSIS VISUALIZATION")
    print("=" * 70)
    
    fs = 44100
    duration = 1
    t = np.arange(0, duration, 1/fs)
    
    # Test signal with multiple frequencies
    test_signal = (
        np.sin(2 * np.pi * 440 * t) +
        0.5 * np.sin(2 * np.pi * 880 * t) +
        0.3 * np.sin(2 * np.pi * 1320 * t)
    )
    
    # Compute FFT
    fft_result = np.fft.fft(test_signal)
    fft_mag = np.abs(fft_result)
    fft_freq = np.fft.fftfreq(len(test_signal), 1/fs)
    
    # Plot
    fig, axes = plt.subplots(2, 1, figsize=(12, 8))
    
    # Time domain
    axes[0].plot(t[:1000], test_signal[:1000])
    axes[0].set_title('Time Domain Signal', fontweight='bold')
    axes[0].set_xlabel('Time (s)')
    axes[0].set_ylabel('Amplitude')
    axes[0].grid(True, alpha=0.3)
    
    # Frequency domain (positive frequencies only)
    pos_mask = fft_freq >= 0
    axes[1].plot(fft_freq[pos_mask][:5000], fft_mag[pos_mask][:5000])
    axes[1].set_title('Frequency Domain (FFT)', fontweight='bold')
    axes[1].set_xlabel('Frequency (Hz)')
    axes[1].set_ylabel('Magnitude')
    axes[1].grid(True, alpha=0.3)
    axes[1].set_xlim([0, 2000])
    
    plt.tight_layout()
    plt.savefig('fft_validation_plot.png', dpi=300)
    print("\n✓ Visualization saved: fft_validation_plot.png")
    plt.show()


if __name__ == '__main__':
    print("\n" + "=" * 70)
    print("RUNNING FFT VALIDATION SUITE")
    print("=" * 70)
    
    results = []
    
    # Run validations
    results.append(validate_parsevals_theorem())
    results.append(validate_spectral_centroid())
    
    # Create visualization
    visualize_fft_analysis()
    
    # Summary
    passed = sum(results)
    total = len(results)
    
    print("\n" + "=" * 70)
    print("VALIDATION SUMMARY")
    print("=" * 70)
    print(f"Passed: {passed}/{total}")
    print(f"Failed: {total - passed}/{total}")
    
    if passed == total:
        print("\n✓ All validations passed!")
    else:
        print(f"\n✗ {total - passed} validation(s) failed")
